cardano-api/src/Cardano/Api/Fees.hs

```diff
@@ -793,68 +794,105 @@ makeTransactionBodyAutoBalance eraInMode systemstart history pparams
     explicitTxFees <- first (const TxBodyErrorByronEraNotSupported) $
                         txFeesExplicitInEra era'
 
-    -- Insert change address and set tx fee to 0
+    -- Make a txbody that we will use for calculating the fees. For the purpose
+    -- of fees we just need to make a txbody of the right size in bytes. We do
+    -- not need the right values for the fee or change output. We use
+    -- "big enough" values for the change output and set so that the CBOR
+    -- encoding size of the tx will be big enough to cover the size of the final
+    -- output and fee. Yes this means this current code will only work for
+    -- final fee and change output of less than around 4000 ada (2^32-1 lovelace)
     txbody1 <- first TxBodyError $ -- TODO: impossible to fail now
                makeTransactionBody txbodycontent1 {
-                 txFee = TxFeeExplicit explicitTxFees 0
+                 txFee  = TxFeeExplicit explicitTxFees $ Lovelace (2^(31 :: Integer)),
+                 txOuts = TxOut changeaddr
+                                (lovelaceToTxOutValue $ Lovelace (2^(31 :: Integer)))
```

Comment:
we need to reserve more space for the change address. it could potentially take up 8 bytes. reserving 4 bytes for the fee is great, though.